### DISCUSSION TITLE
HSS boot from which device

### DISCUSSION THREAD

**Discussion by xiyanni** [polarfire-soc/polarfire-soc-documentation#99] [HSS and U-Boot] [answered]
Does hss do same sequence in default?
Why is any of the boot sources? Is not any priority?
Does user choose the different storage with command or configuration? 
How to achieve the storage dynamically?  
Example:prepare a SD card with image->boot LINUX from SD and copy image to QSPI->remove SD and boot Linux from QSPI.

Below is copy from the link , and the questions also comes from here: https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/software-development/polarfire-soc-software-tool-flow.md#build-the-hss

By default, the sequence is as follows:

The HSS will attempt to initialize the QSPI flash memory device

If no flash memory is connected, the HSS will attempt to initialize the SD card

If no card is inserted, the HSS will attempt to initialize the eMMC

There will then be a count down waiting for a user input on the HSS UART console. If no key is pressed, the HSS will attempt to load a payload from any of the boot sources mentioned above. If a key is pressed the user can enter a service such as the USBDMSC for mounting and programming either the external QSPI flash memory or eMMC.

**Comment by hughbreslin** [comment, marked as answer]
Hi @xiyanni :) To answer your questions:

1. The HSS will always use the default boot sequence that was outlined unless the configuration is changed to remove one of the boot sources, e.g if QSPI is not configured in the HSS build it will not attempt to boot from it or if the boot order / priority is changed by the user.

2. The reasoning behind the ordering is that the removable boot sources are prioritized, QSPI flash might not always be present, the SD card can be removed but the eMMC is always there and is used as a fallback. This is only really intended for development and we expect users to define the boot sources they will use in a final application. Users can select different boot sources using commands The process is described in this [document](https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/boards/mpfs-icicle-kit-es/updating-icicle-kit/updating-icicle-kit-design-and-linux.md#programming-the-linux-image)

3. To achieve the configuration you've described you'll need to re-prioritize the boot sources so the SD card takes priority over the QSPI and the Linux image can then boot up from the SD card as the default memory and copy over to QSPI. This can be done through the [hss_boot_init.c](https://github.com/polarfire-soc/hart-software-services/blob/master/init/hss_boot_init.c#L135) file.
